rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -940,209 +908,149 @@ impl CanisterQueues {
-            }
+        let context = id.context();
+        let remote = match context {
+            Inbound => msg.sender(),
+            Outbound => msg.receiver(),
+        };
+        let (input_queue, output_queue) = self.canister_queues.get_mut(&remote).unwrap();
+        let (queue, reverse_queue) = match context {
+            Inbound => (input_queue, output_queue),
+            Outbound => (output_queue, input_queue),
+        };
+
+        // Ensure that the first reference in a queue is never stale: if we dropped the
+        // message at the head of a queue, advance to the first non-stale reference.
+        //
+        // Defensive check, reference may have already been popped by an earlier
+        // `on_message_dropped()` call if multiple messages got dropped at once.
+        if queue.peek() == Some(&CanisterQueueItem::Reference(id)) {
+            queue.pop();
+            queue.pop_while(|item| self.pool.get(item.id()).is_none());
         }
 
-        debug_assert!(self.stats_ok());
-        debug_assert_eq!(Ok(()), self.schedules_ok(own_canister_id, local_canisters));
+        // Generate reject response, iff this was an outbound request. Or release the
+        // response slot iff this was an inbound request.
+        let request = match (context, msg) {
+            // Outbound request: produce a `SYS_TRANSIENT` timeout response.
+            (Outbound, RequestOrResponse::Request(request)) => request,
+
+            // Inbound request: release the outbound response slot and return.
+            (Inbound, RequestOrResponse::Request(request)) => {
+                reverse_queue.release_reserved_response_slot();
+                self.queue_stats.on_drop_input_request(request);
+                return;
+            }
 
-        timed_out_requests_count
+            // Inbound or outbound response, nothing left to do.
+            //
+            // TODO(MR-603): Recall the `Id` -> `CallbackId` of shed inbound responses and
+            // generate a reject response on the fly when the respective `Id` is popped.
+            (_, RequestOrResponse::Response(_)) => return,
+        };
+        let response = generate_timeout_response(request);
+        let destination = &request.receiver;
+        let (input_queue, _) = self.canister_queues.get_mut(destination).unwrap();
+
+        // Update stats for the generated response.
+        self.queue_stats.on_push_response(&response, Inbound);
+
+        let id = self.pool.insert_inbound(response.into());
+        input_queue.push_response(id);
+
+        // If the input queue is not already in an input schedule, add it.
+        if input_queue.len() == 1 && self.input_schedule_canisters.insert(remote) {
+            if &remote == own_canister_id || local_canisters.contains_key(&remote) {
+                self.local_subnet_input_schedule.push_back(remote)
+            } else {
+                self.remote_subnet_input_schedule.push_back(remote)
+            }
+        }
```

Comment:
Unless I'm missing something, this block is only executed if we are in the `(Outbound, RequestOrResponse::Request(request))` case. Should we move it up?

Reply:
We could, but it would be indented to kingdom come. It may be easier to read, though. I'll give it a try.

Reply:
If the indentation is the problem also a comment indicating that we only continue here if this is an outbound request might do.